belt-block/src/lib.rs

```diff
@@ -0,0 +1,223 @@
+    // 2.2) c <- c xor G21(a+K[7i-5])
+    data[c as usize] ^= g21(data[d as usize].wrapping_add(key[(7 * i as usize - 5 - 1) % 8]));
+    // 2.3) a <- a - G13(a+K[7i-4])
+    data[a as usize] = data[a as usize].wrapping_sub(g13(
+        data[b as usize].wrapping_add(key[(7 * i as usize - 4 - 1) % 8])
+    ));
+    // 2.4) e <- G21(b+c+K[7i-3])+<i>_32
+    let e = g21(data[b as usize]
+        .wrapping_add(data[c as usize])
+        .wrapping_add(key[(7 * i as usize - 3 - 1) % 8]))
+        ^ i;
+    // 2.5) b <- b+e
+    data[b as usize] = data[b as usize].wrapping_add(e);
+    // 2.6) c <- c-e
+    data[c as usize] = data[c as usize].wrapping_sub(e);
+    // 2.7) d <- d xor G13(c+K[7i-2])
+    data[d as usize] = data[d as usize].wrapping_add(g13(
+        data[c as usize].wrapping_add(key[(7 * i as usize - 2 - 1) % 8])
+    ));
+    // 2.8) b <- b xor G21(a + K[7i-1])
+    data[b as usize] ^= g21(data[a as usize].wrapping_add(key[(7 * i as usize - 1 - 1) % 8]));
+    // 2.9) c <- c xor G5(d+K[7i])
+    data[c as usize] ^= g5(data[d as usize].wrapping_add(key[(7 * i as usize - 1) % 8]));
+}
+
+#[inline]
+#[allow(clippy::many_single_char_names)]
+/// Tact substitution, stands for 2.1-2.9 steps 6.1.4 algorithm (Block decryption) STB 34.101.31-2011
+fn tact_dec(i: u32, a: u32, b: u32, c: u32, d: u32, data: &mut [u32], key: &[u32]) {
+    // 2.1) b <- b xor G5(a+K[7i])
+    data[b as usize] ^= g5(data[a as usize].wrapping_add(key[(7 * i as usize - 1) % 8]));
+    // 2.2) c <- c xor G21(a+K[7i-1])
+    data[c as usize] ^= g21(data[d as usize].wrapping_add(key[(7 * i as usize - 1 - 1) % 8]));
+    // 2.3) a <- a - G13(a+K[7i-2])
+    data[a as usize] = data[a as usize].wrapping_sub(g13(
+        data[b as usize].wrapping_add(key[(7 * i as usize - 2 - 1) % 8])
+    ));
+    // 2.4) e <- G21(b+c+K[7i-3])+<i>_32
+    let e = g21(data[b as usize]
+        .wrapping_add(data[c as usize])
+        .wrapping_add(key[(7 * i as usize - 3 - 1) % 8]))
+        ^ i;
+    // 2.5) b <- b+e
+    data[b as usize] = data[b as usize].wrapping_add(e);
+    // 2.6) c <- c-e
+    data[c as usize] = data[c as usize].wrapping_sub(e);
+    // 2.7) d <- d xor G13(c+K[7i-4])
+    data[d as usize] = data[d as usize].wrapping_add(g13(
+        data[c as usize].wrapping_add(key[(7 * i as usize - 4 - 1) % 8])
+    ));
+    // 2.8) b <- b xor G21(a + K[7i-5])
+    data[b as usize] ^= g21(data[a as usize].wrapping_add(key[(7 * i as usize - 5 - 1) % 8]));
+    // 2.9) c <- c xor G5(d+K[7i-6])
+    data[c as usize] ^= g5(data[d as usize].wrapping_add(key[(7 * i as usize - 6 - 1) % 8]));
+}
+
+impl BeltBlock {
+    #[inline]
+    pub(crate) fn encrypt(&self, data: &mut [u8; 16]) {
+        // Represent data as 4u32
+        let data_u32 = unsafe {
+            &mut *slice_from_raw_parts_mut(
+                data.as_mut_ptr() as *mut u32,
+                data.len() / core::mem::size_of::<u32>(),
+            )
```

Comment:
This line is incorrect. It will produce incorrect results on big-endian targets. You can use the [`sm4` code](https://github.com/RustCrypto/block-ciphers/blob/7de364ede310f1ea7080c99ddf1138aeb47f9a69/sm4/src/lib.rs#L132-L137) for reference,

Reply:
Instead of manual inlining you may use `zip`-ing loops.

Reply:
Also, I forget to mention, but this code in addition to the endianness issue had problem with alignment. `data` may not have 4-byte aligment required by `u32`. In practice it can cause CPU exception raised by aligned load/store being used on an unaligned pointer.